test/common/init/target_impl_test.cc

```diff
@@ -60,6 +62,110 @@ TEST(InitTargetImplTest, ReadyWhenWatcherUnavailable) {
   EXPECT_FALSE(target.ready());
 }
 
+TEST(SharedTargetImplTest, OnSyncManagers) {
+  ManagerImpl m1("m1");
+  ManagerImpl m2("m2");
+  int counter = 0;
+  SharedTargetImpl t("test", [&counter]() mutable { counter++; });
+  m1.add(t);
+  EXPECT_EQ(0, counter);
+  m2.add(t);
+  EXPECT_EQ(0, counter);
+  EXPECT_EQ(Manager::State::Uninitialized, m1.state());
+  EXPECT_EQ(Manager::State::Uninitialized, m2.state());
+  m1.initialize(WatcherImpl("watcher", []() {}));
+  m2.initialize(WatcherImpl("watcher", []() {}));
+  EXPECT_LE(1, counter);
+  EXPECT_EQ(Manager::State::Initializing, m1.state());
+  EXPECT_EQ(Manager::State::Initializing, m2.state());
+
+  t.ready();
+  EXPECT_EQ(Manager::State::Initialized, m1.state());
+  EXPECT_EQ(Manager::State::Initialized, m1.state());
+}
+
+// Two managers initialize the same target at their own interests.
+TEST(SharedTargetImplTest, ConcurrentManagerInitialization) {
+  ManagerImpl m1("m1");
+  ManagerImpl m2("m2");
+  int counter = 0;
+  SharedTargetImpl t("test", [&counter]() mutable { counter++; });
+  m1.add(t);
+  EXPECT_EQ(0, counter);
+  m2.add(t);
+  EXPECT_EQ(0, counter);
+  EXPECT_EQ(Manager::State::Uninitialized, m1.state());
+  EXPECT_EQ(Manager::State::Uninitialized, m2.state());
+  m1.initialize(WatcherImpl("watcher", []() {}));
+  EXPECT_EQ(1, counter);
+  EXPECT_EQ(Manager::State::Initializing, m1.state());
+  EXPECT_EQ(Manager::State::Uninitialized, m2.state());
+  t.ready();
+  EXPECT_EQ(Manager::State::Initialized, m1.state());
+  EXPECT_EQ(Manager::State::Uninitialized, m2.state());
+  m2.initialize(WatcherImpl("watcher", []() {}));
+  EXPECT_EQ(Manager::State::Initialized, m2.state());
+  EXPECT_EQ(1, counter) << "target init function should be invoked only once";
+}
+
+// One manager initialized the target, followed by another manager adding the target.
+TEST(SharedTargetImplTest, OnLateManagers) {
+  ManagerImpl m1("m1");
+  ManagerImpl m2("m2");
+  int counter = 0;
+  SharedTargetImpl t("test", [&counter]() mutable { counter++; });
+  m1.add(t);
+  EXPECT_EQ(0, counter);
+  EXPECT_EQ(Manager::State::Uninitialized, m1.state());
+  m1.initialize(WatcherImpl("watcher", []() {}));
+  EXPECT_EQ(1, counter);
+  EXPECT_EQ(Manager::State::Initializing, m1.state());
+  t.ready();
+  EXPECT_EQ(Manager::State::Initialized, m1.state());
+
+  EXPECT_EQ(Manager::State::Uninitialized, m2.state());
+  m2.initialize(WatcherImpl("watcher", []() {}));
+  EXPECT_EQ(Manager::State::Initialized, m2.state());
+}
+
+TEST(SharedTargetImplTest, GoneTargetIsInitialized) {
```

Comment:
See how `InitializeWhenUnavailable` is implemented above; I think you just need to replicate that.